File: C++/2-Medium/00256._粉刷房子/00256._粉刷房子.cpp

#include <iostream>
#include <vector>
using namespace std;

// 动态规划解决粉刷房子问题的函数
int minCost(vector<vector<int>>& costs) {
    if (costs.empty()) return 0;
    int n = costs.size();
    vector<vector<int>> dp(n, vector<int>(3, 0));
    // 初始化第一个房子的成本
    for (int j = 0; j < 3; ++j) {
        dp[0][j] = costs[0][j];
    }
    // 动态规划计算每个房子的最低成本
    for (int i = 1; i < n; ++i) {
        dp[i][0] = min(dp[i-1][1], dp[i-1][2]) + costs[i][0];
        dp[i][1] = min(dp[i-1][0], dp[i-1][2]) + costs[i][1];
        dp[i][2] = min(dp[i-1][0], dp[i-1][1]) + costs[i][2];
    }
    // 返回最后一个房子的最低成本
    return min(min(dp[n-1][0], dp[n-1][1]), dp[n-1][2]);
}

int main() {
    // 示例
    vector<vector<int>> costs = {{17, 2, 17}, {16, 16, 5}, {14, 3, 19}};
    cout << "Minimum cost to paint all houses: " << minCost(costs) << endl;
    return 0;
}

File: Python/2-Medium/00256._粉刷房子/00256._粉刷房子.py
from typing import List

def minCost(costs: List[List[int]]) -> int:
    """
    计算粉刷所有房子的最低成本。
    
    参数:
    costs: 一个二维列表，代表每个房子粉刷成每种颜色的成本。
    
    返回:
    int: 所有房子粉刷的最低成本。
    
    示例:
    >>> minCost([[17,2,17],[16,16,5],[14,3,19]])
    10
    """
    if not costs:
        return 0  # 如果没有房子，成本为0
    
    n = len(costs)
    dp = costs.copy()  # 初始化dp数组
    
    # 动态规划计算最低成本
    for i in range(1, n):
        # 粉刷第i个房子为红色的最低成本
        dp[i][0] = min(dp[i-1][1], dp[i-1][2]) + costs[i][0]
        # 粉刷第i个房子为蓝色的最低成本
        dp[i][1] = min(dp[i-1][0], dp[i-1][2]) + costs[i][1]
        # 粉刷第i个房子为绿色的最低成本
        dp[i][2] = min(dp[i-1][0], dp[i-1][1]) + costs[i][2]
    
    # 返回所有房子粉刷的最低成本
    return min(dp[-1])

# 测试代码
test_costs = [[17,2,17],[16,16,5],[14,3,19]]
print(minCost(test_costs))  # 应输出最低成本